server/etcdserver/api/etcdhttp/readiness.go

```diff
@@ -0,0 +1,202 @@
+
+func InstallHealthEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	InstallLivezEndpoints(lg, mux, server)
+	InstallReayzEndpoints(lg, mux, server)
+}
+
+type HealthCheck func(ctx context.Context) error
+
+type CheckRegistry struct {
+	path string
+
+	checks map[string]HealthCheck
+	// the list of checkNames are needed to ensure the order of checks.
+	checkNames []string
+}
+
+func InstallLivezEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/livez", checks: make(map[string]HealthCheck)}
+	reg.Register("serializable_read", checkSerializableRead(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func InstallReayzEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/readyz", checks: make(map[string]HealthCheck)}
+	reg.Register("data_corruption", checkAlarm(server, etcdserverpb.AlarmType_CORRUPT))
+	reg.Register("serializable_read", checkSerializableRead(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func (reg *CheckRegistry) Register(name string, check HealthCheck) {
+	reg.checkNames = append(reg.checkNames, name)
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) InstallHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
+	lg.Info(fmt.Sprintf("Installing %s http endpoint.", reg.path))
+	mux.Handle(reg.path, reg.handleRootHealth(lg))
+	for checkName, check := range reg.checks {
+		mux.Handle(fmt.Sprintf("%s/%v", reg.path, checkName), adaptCheckToHandler(check))
+	}
+}
+
+// handleRootHealth returns an http.HandlerFunc that serves the provided checks.
+func (reg *CheckRegistry) handleRootHealth(lg *zap.Logger) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// extracts the health check names to be excludeList from the query param
+		excludeList, _ := r.URL.Query()["exclude"]
+		// extracts the health check names to be allowList from the query param
+		allowList, _ := r.URL.Query()["allowlist"]
+
+		if len(excludeList) > 0 && len(allowList) > 0 {
+			http.Error(w, fmt.Sprintf("do not expect both allowlist: %v and exclude list: %v to be both specified in %s.", allowList, excludeList, reg.path), http.StatusBadRequest)
+			return
+		}
+		excluded := listToStringSet(excludeList)
+		included := listToStringSet(allowList)
+		failedChecks := []string{}
+		var individualCheckOutput bytes.Buffer
+		for _, checkName := range reg.checkNames {
+			if len(allowList) > 0 {
+				if _, found := included[checkName]; !found {
+					fmt.Fprintf(&individualCheckOutput, "[+]%s not included: ok\n", checkName)
+					continue
+				}
+				delete(included, checkName)
+			} else {
+				// no-op the check if we've specified we want to exclude the check
+				if _, found := excluded[checkName]; found {
+					delete(excluded, checkName)
+					fmt.Fprintf(&individualCheckOutput, "[+]%s excluded: ok\n", checkName)
+					continue
+				}
+			}
+			check := reg.checks[checkName]
+			if err := check(r.Context()); err != nil {
```

Comment:
IMHO, code readability matters more than time complexity. So instead of mixing the exclude and check logic, I would suggest

1. get a list of checks given (reg.checkNames, excluded and included).
2. return http bad request to client if the intended (excluded/included) checks are not registered. So it forces administrator to configure it correctly because etcd will be in a crash back off state. 
3. run each check individually.

Reply:
done

Reply:
Sync'd up with @siyuanfoundation. 

“Failing health check if exclude query contains some non registered check” would be a blocker for us to remove a check. So we could switch to log an error with warn log level. Downside is admin needs to check the warn log to figure out if the prober setting is correct instead. 

Version compatibility outweighs forcing administrator to configure it correctly.. 

ref. https://github.com/kubernetes/kubernetes/pull/70676#discussion_r231610436